packages/falso/src/lib/core/core.ts

```diff
@@ -1,40 +1,116 @@
 import { random } from '../random';
+import { primitiveValueIsUnique } from './unique-validators';
 
 export interface FakeOptions {
   length?: number;
+  priority?: 'length' | 'unique';
   locale?: any | string[];
 }
 
 export type markRequired<Type, Key extends keyof Type> = Type & {
   [Property in Key]-?: Type[Property];
 };
 
+export interface FakeConfig<T> {
+  uniqueComparer: (
+    item: T,
+    items: T[],
+    comparisonKeys?: (keyof T)[]
+  ) => boolean | ((item: T, items: T[]) => boolean);
+  comparisonKeys?: (keyof T)[];
+}
+
 export type Return<T, O extends FakeOptions> = [O] extends [never]
   ? T
   : O['length'] extends number
   ? T[]
   : T;
 
-type FactoryFunction<T> = (i: number) => T;
+export type FactoryFunction<T> = (i: number) => T;
 
 export function fake<T, Options extends FakeOptions>(
   data: Readonly<T[]> | FactoryFunction<T>,
-  options?: Options
+  options?: Options,
+  config: FakeConfig<T> = {
+    uniqueComparer: primitiveValueIsUnique,
+  }
 ): Return<T, Options> {
-  const dataSource = Array.isArray(data)
-    ? () => randElement(data)
-    : (data as FactoryFunction<T>);
+  if (options?.length === 0) {
+    return [] as any;
+  }
+
+  if (Array.isArray(data)) {
+    return fakeFromArray(data, options) as any;
+  }
+
+  return fakeFromFunction(data as FactoryFunction<T>, config, options) as any;
+}
+
+export function fakeFromFunction<T, Options extends FakeOptions>(
+  data: FactoryFunction<T>,
+  config: FakeConfig<T>,
+  options?: Options
+): T | T[] {
+  if (!options?.length) {
+    return data(0);
+  }
+
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, (_, index) => data(0));
+  }
+
+  const items: T[] = [];
+
+  let attempts = 0;
+  const maxAttempts = options.length * 2;
+
+  while (items.length < options.length && attempts < maxAttempts) {
+    const item = data(0);
+
+    if (config.uniqueComparer(item, items, config.comparisonKeys)) {
+      items.push(item);
+    }
+
+    attempts++;
+  }
+
+  return items;
+}
+
+export function fakeFromArray<T, Options extends FakeOptions>(
+  data: T[],
+  options?: Options
+): T | T[] {
+  if (!options?.length) {
+    return randElement(data);
+  }
+
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, () => randElement(data));
+  }
+
+  const clonedData: T[] = JSON.parse(JSON.stringify(data));
+  const newArray: T[] = [];
+
+  while (clonedData.length && newArray.length !== options.length) {
+    const randomIndex = getRandomInRange({
```

Comment:
There is a hidden assumption here that the array you are given holds unique values.
Please change the code to support non-unique arrays.

Reply:
@shaharkazaz - You are correct. How did I not spot that :/ 

I have reworked it to shuffle the data array and then pick the first n items from it. A positive of doing it this way is that the user will receive an array with a length as close as possible to their desired length. The issue is that this will be more expensive for large data sets.

```ts 
export function fakeFromArray<T, Options extends FakeOptions>(
  data: T[],
  options?: Options
): T | T[] {
  if (options?.length === 0) {
    return [];
  }

  if (!options?.length) {
    return randElement(data);
  }

  const priority = options?.priority ?? 'length';

  if (priority === 'length') {
    return Array.from({ length: options.length }, () => randElement(data));
  }

  let clonedData: T[] = JSON.parse(JSON.stringify(data));
  // Randomise order of data
  clonedData = shuffleArray<T>(clonedData);

  if (options.length >= clonedData.length) {
    return clonedData;
  }

  return clonedData.slice(0, length);
}

function shuffleArray<T>(array: T[]): T[] {
  for (let i = array.length - 1; i > 0; i--) {
    const randomIndex = Math.floor(random() * (i + 1));

    [
      array[i],
      array[randomIndex],
    ] = [
      array[randomIndex],
      array[i],
    ];
  }

  return array;
}
```

What are your thoughts?